include/common/graphic-helpers.h

```diff
@@ -9,29 +9,49 @@ struct wlr_scene_tree;
 struct wlr_scene_rect;
 struct wlr_fbox;
 
+struct border_rect {
+	struct wlr_scene_tree *tree;
+	int border_width; /* read-only */
+
+	/* Private */
+	struct wlr_scene_rect *top;
+	struct wlr_scene_rect *bottom;
+	struct wlr_scene_rect *left;
+	struct wlr_scene_rect *right;
+	struct wlr_scene_rect *bg;
+
+	struct wl_listener destroy;
```

Comment:
I can live with it, but there is something that seems a bit unbalanced about having nested trees and destroy-listeners in the case of the multi_rect case.

Suggest considering breaking out the code below into a separate struct:

```
int border_width;
struct wlr_scene_rect *top;
struct wlr_scene_rect *bottom;
struct wlr_scene_rect *left;
struct wlr_scene_rect *right;
struct wlr_scene_rect *bg;
```

...leaving the (currently named) `border_rect` with the `tree`, a `*border_rect` and a `destroy`.

Reply:
Just thinking aloud...

```C
struct rect {
	int border_thickness;
	struct wlr_scene_rect *top;
	struct wlr_scene_rect *bottom;
	struct wlr_scene_rect *left;
	struct wlr_scene_rect *right;
	struct wlr_scene_rect *fill;
}

struct scene_object {
	struct wlr_scene_tree *tree;
	struct wl_listener destroy;
};

struct lab_scene_rect {
	struct scene_object base;
	struct rect rect;
};

struct lab_scene_multi_rect {
	struct scene_object base;
	struct rect rects[3];
};
```

Reply:
> Yeah I had a similar feeling.
> 
> I opened #2809 as an alternative solution. What do you think of it?

That looks a much nicer interface. Great work!